platform-dependent-components/problem-solver/cxx/sections-module/agent/get_decomposition_agent.cpp

```diff
@@ -17,75 +16,80 @@ using namespace utils;
 
 namespace sectionsModule
 {
-SC_AGENT_IMPLEMENTATION(GetDecompositionAgent)
+ScResult GetDecompositionAgent::DoProgram(ScActionInitiatedEvent const & event, ScAction & action)
 {
-  ScAddr const questionNode = m_memoryCtx.GetEdgeTarget(edgeAddr);
+  ScAddr const questionNode = m_context.GetEdgeTarget(event.GetArc());
   if (!CheckActionClass(questionNode))
-    return SC_RESULT_OK;
+    return action.FinishSuccessfully();
 
   SC_LOG_INFO("GetDecompositionAgent started");
 
   size_t level = 1;
-  ScAddr const subjDomainAddr =
-      IteratorUtils::getAnyByOutRelation(&m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_1);
-  ScAddr const levelAddr =
-      IteratorUtils::getAnyByOutRelation(&m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_2);
-  ScAddr const langAddr =
-      IteratorUtils::getAnyByOutRelation(&m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_3);
-  ScAddr decompositionAddr =
-      IteratorUtils::getAnyByOutRelation(&m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_4);
-
-  if (!m_memoryCtx.IsElement(subjDomainAddr))
+
+  std::tuple<ScAddr, ScAddr, ScAddr, ScAddr> const Tresult = action.GetArguments<4>();
+  ScAddr const subjDomainAddr = std::get<0>(Tresult);
+  ScAddr const levelAddr = std::get<1>(Tresult);
+  ScAddr const langAddr = std::get<2>(Tresult);
+  ScAddr decompositionAddr = std::get<3>(Tresult);
```

Comment:
```suggestion
  auto const [subjDomainAddr, levelAddr, langAddr, decompositionAddr] = action.GetArguments<4>();
```

Reply:
it's bad that decompositionAddr is not const because you won't be able to use what I  suggested, but something similar can be used